util/http/http.go

```diff
@@ -2,27 +2,82 @@ package http
 
 import (
 	"fmt"
+	"math"
 	"net/http"
 	"net/http/httputil"
 	"strings"
 
 	log "github.com/sirupsen/logrus"
 )
 
+// max number of chunks a cookie can be broken into. To be compatible with
+// widest range of browsers, we shouldn't create more than 30 cookies per domain
+const maxNumber = 5
+const maxLength = 4093
+
 // MakeCookieMetadata generates a string representing a Web cookie.  Yum!
-func MakeCookieMetadata(key, value string, flags ...string) (string, error) {
-	components := []string{
-		fmt.Sprintf("%s=%s", key, value),
+func MakeCookieMetadata(key, value string, flags ...string) ([]string, error) {
+	attributes := strings.Join(flags, "; ")
+
+	// cookie: name=value; attributes and key: key-(i) e.g. argocd.token-0
+	maxValueLength := maxValueLength(key, attributes)
+	numberOfCookies := int(math.Ceil(float64(len(value)) / float64(maxValueLength)))
+	if numberOfCookies > maxNumber {
+		return nil, fmt.Errorf("invalid cookie value, at %d long it is longer than the max length of %d", len(value), maxValueLength*maxNumber)
+	}
+
+	return splitCookie(key, value, attributes), nil
+}
+
+// browser has limit on size of cookie, currently 4kb. In order to
+// support cookies longer than 4kb, we split cookie into multiple 4kb chunks.
+func splitCookie(key, value, attributes string) []string {
+	var cookies []string
+	valueLength := len(value)
+
+	// cookie: name=value; attributes and key: key-(i) e.g. argocd.token-0
+	maxValueLength := maxValueLength(key, attributes)
+	var end int
+	for i, j := 0, 0; i < valueLength; i, j = i+maxValueLength, j+1 {
+		end = i + maxValueLength
+		if end > valueLength {
+			end = valueLength
+		}
+		if attributes == "" {
+			cookies = append(cookies, fmt.Sprintf("%s-%d=%s", key, j, value[i:end]))
+		} else {
+			cookies = append(cookies, fmt.Sprintf("%s-%d=%s; %s", key, j, value[i:end], attributes))
+		}
 	}
-	components = append(components, flags...)
-	header := strings.Join(components, "; ")
+	return cookies
+}
+
+// JoinCookies combines chunks of cookie based on key as prefix.
+// It returns cookie value as string. cookieString is of format
+// key1=value1; key2=value2; key3=value3
+func JoinCookies(key string, cookieString string) string {
+	cookies := make(map[string]string)
+	for _, cookie := range strings.Split(cookieString, ";") {
```

Comment:
Reproduce an issue:

* User is part of many groups. Login to Argo CD and got 2 cookie chunks.
* Remove all groups and login again: small cookie was generated and only first chunk got replaced. So browser has two cookies and second if leftover from first login
* User unable to login anymore

I really want to avoid asking users to clean cookies and try to login again. I would suggest that if cookie is chunked then first part must include information about number of chunks. E.g. `parts=<count>:<token>`

Reply:
It's a good idea to add number of chunks in first part. I have added the changes. 

To point out the above issue, I have added code to make sure to clear all cookies with prefix `argocd.token`